src/main/java/com/github/firmwehr/gentle/semantic/analysis/MainMethodLookupVisitor.java

```diff
@@ -0,0 +1,85 @@
+package com.github.firmwehr.gentle.semantic.analysis;
+
+import com.github.firmwehr.gentle.semantic.SemanticException;
+import com.github.firmwehr.gentle.semantic.Visitor;
+import com.github.firmwehr.gentle.semantic.ast.LocalVariableDeclaration;
+import com.github.firmwehr.gentle.semantic.ast.SMethod;
+import com.github.firmwehr.gentle.semantic.ast.expression.SLocalVariableExpression;
+import com.github.firmwehr.gentle.semantic.ast.type.SVoidType;
+import com.github.firmwehr.gentle.source.Source;
+
+import java.util.Optional;
+
+/**
+ * Finds the main method, ensures it is globally unique and exists and:
+ * <ul>
+ *     <li>Is named main</li>
+ *     <li>Has a single parameter of type String[]</li>
+ *     <li>Has a void return type</li>
+ *     <li>Is static</li>
+ *     <li>The parameter is never used in the method</li>
+ * </ul>
+ */
+public class MainMethodLookupVisitor implements Visitor<Optional<Void>> {
+	private final Source source;
+
+	private Optional<SMethod> foundMainMethod = Optional.empty();
+	private Optional<LocalVariableDeclaration> mainMethodParameter = Optional.empty();
+
+	public MainMethodLookupVisitor(Source source) {
+		this.source = source;
+	}
+
+	@Override
+	public Optional<Void> defaultReturnValue() {
+		return Optional.empty();
+	}
+
+	@Override
+	public Optional<Void> visit(SMethod method) throws SemanticException {
+		if (!method.isStatic()) {
+			return Optional.empty();
+		}
+		if (this.foundMainMethod.isPresent()) {
+			throw new SemanticException(source, method.name().sourceSpan(), "duplicate main method",
+				foundMainMethod.get().name().sourceSpan(), "already defined here");
+		}
+		if (!method.name().ident().equals("main")) {
+			throw new SemanticException(source, method.name().sourceSpan(), "main method must be named 'main'");
+		}
+
+		if (!(method.returnType() instanceof SVoidType)) {
+			throw new IllegalArgumentException("The main method must have a void return type");
+		}
+		if (method.parameters().size() != 1) {
+			throw new IllegalArgumentException("The main method must have exactly one parameter");
```

Comment:
Shouldn't those be `SemanticException`s too?

Reply:
These cases are already prevented by the parser, so these conditions should always be false. This just double-checks them.